### PR TITLE
[Opt] CrossEntropy: broadcast optimization + FP32 dedicated path + tail padding

### PR DESCRIPTION
**Pull request by zwj-sep**
## 改动内容

基于 `examples/cross_entropy_loss/example_cross_entro.py` 进行性能优化，直接覆盖原文件，包含 3 个优化点 + 6 个 reviewer 修复。

### 优化点

**1. 尾数处理（pad_value）**
- `T.copy` 加 `pad_value=-inf`，C 不整除 block_C 时防越界

**2. broadcast 替代逐行循环**
- 原版：`for n_idx: T.tile.sub(x[n_idx,:], max[n_idx])` 64 次串行
- 优化：`broadcast(max_2d, max, axis=1) + T.tile.sub(x, max_2d)` 1 次整体运算
- float16 bc=128: 8.1ms → 3.7ms（2.19x）

**3. FP32 专用 kernel**
- float32 输入时去掉 `x_ub` buffer（省 64KB），直接搬到 `x_32`
- block_C 可开到 192
- float32 bc=192: 8.1ms → 2.9ms（2.79x）

### Reviewer 修复

1. **broadcast axis=1**：方形 block 时自动推断选错维度，59/64 mismatch
2. **y_dtype 去掉 int64**：AscendC 不支持 int64 Adds，编译失败
3. **数据竞争**：modulo 改线性 bn + `if bn < n_2_num` 守卫
4. **FP32 多余 buffer**：去掉 `l_n`，直接 `T.copy(l_n_32, loss)`
5. **循环外 broadcast**：`prev_max`/`prev_sum` 移到第二遍循环外，减少 2048 次重复 broadcast
6. **测试扩展**：19→23 用例，加方形 block / c_num=1 / float32 尾数 / 不同 batch

### 内部性能对比（N=4, C=131072）

**float16:**
| 版本 | block_C | 耗时 | vs 原版 |
|------|---------|------|--------|
| 原版（逐行） | 128 | 8.1ms | 1.0x |
| 原版（逐行） | 256 | 4.6ms | 1.76x |
| +broadcast | 128 | 3.9ms | 2.08x |
| +broadcast+axis+logsum_2d | 128 | 3.7ms | 2.19x |

**float32:**
| 版本 | block_C | 耗时 | vs 原版 |
|------|---------|------|--------|
| 原版（逐行） | 128 | 8.1ms | 1.0x |
| +broadcast+FP32专用 | 128 | 3.5ms | 2.31x |
| +broadcast+FP32专用 | 192 | 2.9ms | 2.79x |

### 性能测试（msprof 纯 Kernel 耗时 + 端到端总耗时）

6 组 shape × 3 种 dtype，block_N=128。

- **TL Kernel (msprof)**：纯 Kernel 设备执行时间，`msprof` `op_statistic` CSV 的 `Min Time(us)`（稳态值，排除冷启动）。
- **TL E2E / CANN E2E**：端到端延迟，`tilelang.profiler.do_bench` 测量（含 L2 cache flush + kernel launch + 执行 + sync）。
- **Speedup**：`CANN_E2E / TL_E2E`（> 1 表示 TileLang 更快）。

| Shape (N, C) | dtype | block | TL Kernel (us) | TL E2E (ms) | CANN E2E (ms) | Speedup |
|---------------|-------|-------|----------------|-------------|---------------|---------|
| (4, 128256)   | fp16  | 128×128 | 3589.21 | 3.6692 | 0.0292 | 0.01x |
| (4, 128256)   | bf16  | 128×128 | 3630.69 | 3.7140 | 0.0294 | 0.01x |
| (4, 128256)   | fp32  | 128×192 | 2784.02 | 2.8927 | 0.0367 | 0.01x |
| (8, 128256)   | fp16  | 128×128 | 3609.65 | 3.6931 | 0.0299 | 0.01x |
| (8, 128256)   | bf16  | 128×128 | 3652.31 | 3.7450 | 0.0303 | 0.01x |
| (8, 128256)   | fp32  | 128×192 | 2797.18 | 2.9208 | 0.0377 | 0.01x |
| (32, 128256)  | fp16  | 128×128 | 3663.19 | 3.7897 | 0.0428 | 0.01x |
| (32, 128256)  | bf16  | 128×128 | 3708.91 | 3.8305 | 0.0430 | 0.01x |
| (32, 128256)  | fp32  | 128×192 | 2985.26 | 3.1287 | 0.0575 | 0.02x |
| (1024, 1024)  | fp16  | 128×128 | 35.24   | 0.0889 | 0.0571 | 0.64x |
| (1024, 1024)  | bf16  | 128×128 | 35.54   | 0.0803 | 0.0523 | 0.65x |
| (1024, 1024)  | fp32  | 128×192 | 33.48   | 0.0815 | 0.0507 | 0.62x |
| (4096, 1024)  | fp16  | 128×128 | 70.24   | 0.0881 | 0.1668 | 1.89x |
| (4096, 1024)  | bf16  | 128×128 | 70.96   | 0.0824 | 0.1661 | 2.02x |
| (4096, 1024)  | fp32  | 128×192 | 65.92   | 0.0907 | 0.1638 | 1.80x |
| (4, 131072)   | fp16  | 128×128 | 3641.97 | 3.7130 | 0.0299 | 0.01x |
| (4, 131072)   | bf16  | 128×128 | 3687.63 | 3.7554 | 0.0298 | 0.01x |
| (4, 131072)   | fp32  | 128×192 | 2840.18 | 2.9499 | 0.0379 | 0.01x |

**关键发现：**

1. **甜点区（N=4096, C=1024）**：TileLang 端到端比 CANN 快 **1.8–2.0×**，纯 Kernel 65–71μs vs CANN 端到端 164–167μs。
2. **均衡 shape（N=1024, C=1024）**：纯 Kernel 33–36μs 快于 CANN 端到端 51–57μs，但 TL 端到端 80–89μs 因 do_bench 的 256MB L2 flush 开销占主导。
3. **小 N 大 C（N=4–32, C=128256/131072）**：TL Kernel 2.8–3.7ms vs CANN ~30μs，约 100× 慢。N=4→32 时间几乎不变——全塞进一个 block_N=128，瓶颈在 c_num 串行循环。
4. **fp32 专用路径全面最快**：block_C=192 比 fp16/bf16 的 128 快 20–25%。

### 已知限制与后续优化

当前 kernel 的 tiling 策略以 `block_N × block_C` 二维 block 为单位处理，在小 N 大 C 场景下（如 N=4, C=131072），`block_N=128` 导致大量 padding 浪费，且 `c_num=1024` 次串行循环使访存带宽成为瓶颈，性能从 tiling 设计上就存在问题。

后续计划增加一个专门处理 small-N/large-C 的 kernel 分支，按 C 维度切分并行度，减少串行循环次数，提升该场景下的性能。

### 测试覆盖（23 用例）

- dtype: float16 / float32 / bfloat16
- block_C: 16 / 32 / 64 / 128 / 192
- block_N: 16 / 32 / 64 / 128
- 方形 block / c_num=1 / 尾数处理 / batch 4~1024

### CODE REVIEW

File: examples/cross_entropy_loss/cross_entropy.py [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+                    # T.copy(x[bn * block_N_2, bc * block_C], x_ub)
+                    T.copy(x[bn * block_N_2, bc * block_C], x_ub, pad_value=-T.infinity(CAL_DTYPE))
+                    cast_or_copy(x_32, x_ub, CAST_MODE_LOW2HIGH, block_N_2 * block_C)
+
+                    T.reduce_max(x_32, tile_max, dim=-1)
+                    T.tile.max(tile_max, prev_max, tile_max)
+                    T.tile.sub(temp_exp, prev_max, tile_max)
+                    T.tile.exp(temp_exp, temp_exp)
+                    T.tile.mul(temp_exp, prev_sum, temp_exp)
+                    T.tile.broadcast(max_2d, tile_max)
```
Comment (ShareableXue) [inline examples/cross_entropy_loss/cross_entropy.py]: Please specify axis=1 explicitly here. The source shape is [block_N_2] and the destination shape is [block_N_2, block_C]; this kernel needs each per-row scalar to be broadcast across all columns of the corresponding row.
When block_C == block_N_2, the destination is square and the current auto-inference logic selects axis=0 first, causing the values to be broadcast along the wrong dimension. This is reproducible with check_case(64, 64, 32, 16), where 59 out of 64 loss elements mismatch.

The auto‑broadcast inference logic is implemented at this line (see the repo file below).
 https://github.com/tile-ai/tilelang-ascend/blob/feac5e6ab59a2282551b44a1e75f669d41cb641c/tilelang/language/ascend_tile.py#L2346-L2369

Please apply axis=1 to all six equivalent broadcast calls in this file and restore the square-tile regression test.

File: examples/cross_entropy_loss/cross_entropy.py [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+        tl.PassConfigKey.TL_ASCEND_MEMORY_PLANNING: True,
+    },
+)
+def cross_entropy(
+    N: int,
+    C: int,
+    block_N: int,
+    block_C: int,
+    x_dtype: Literal["float16", "bfloat16", "float32"] = "float16",
+    y_dtype: Literal["int32", "int64"] = "int32",
```
Comment (ShareableXue) [inline examples/cross_entropy_loss/cross_entropy.py]: The function signature advertises support for y_dtype="int64", but the implementation applies T.tile.sub to y_ub. This lowers to AscendC::Adds<int64_t>, which is not supported by the current AscendC API, so both the generic and FP32 paths fail during compilation.

Please keep the public API consistent with the actual implementation: you may remove "int64" from the Literal and explicitly support only int32 for now. 

BTW, Since PyTorch cross-entropy targets are commonly torch.int64, this should be treated as an important compatibility issue.

File: examples/cross_entropy_loss/cross_entropy.py [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+
+    @T.prim_func
+    def main(
+        x: T.Tensor([N, C], x_dtype),  # type: ignore
+        y: T.Tensor([N], y_dtype),  # type: ignore
+        loss: T.Tensor([N], x_dtype),  # type: ignore
+        log_prob: T.Tensor([N, C], x_dtype),  # type: ignore
+    ):
+        with T.Kernel(n_num, is_npu=True) as (cid, vid):
+            bn = (cid * VEC_NUM + vid) % n_2_num
```
Comment (ShareableXue) [inline examples/cross_entropy_loss/cross_entropy.py]: The modulo maps surplus vector tasks back to an existing bn, causing multiple tasks to write concurrently to the same loss and log_prob regions.

For example, with N=4, block_N=128, both vector cores process bn=0. With N=18192, 286 tasks are launched for only 285 valid tiles, so the final task writes bn=0 again. Even if the computed values are usually identical, these are still overlapping GM writes that introduce a data race and unnecessary work.

Maybe you can keep a linear task_id and guard tasks whose ID is outside n_2_num, instead of wrapping them with modulo.

such as:
```
bn = cid * VEC_NUM + vid
if bn < n_2_num:
    # kernel body
```

File: examples/cross_entropy_loss/cross_entropy.py [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+                    T.tile.sub(x_32, x_32, max_2d)
+
+                    T.copy(x_32, log_prob[bn * block_N_2, bc * block_C])
+
+                    for n_idx in T.serial(block_N_2):
+                        if y_ub[n_idx] >= 0 and y_ub[n_idx] < block_C:
+                            l_n_32[n_idx] = -x_32[n_idx, y_ub[n_idx]]
+                    T.tile.sub(y_ub, y_ub, block_C)
+
+                T.copy(l_n_32, l_n)
```
Comment (ShareableXue) [inline examples/cross_entropy_loss/cross_entropy.py]: In the FP32 path, l_n_32, l_n, and the output loss all have the same FP32 dtype, and l_n is not used anywhere else. Therefore, T.copy(l_n_32, l_n) is an extra UB-to-UB copy and does not perform any type conversion.

Please consider removing the l_n buffer and writing the result directly with T.copy(l_n_32, loss[bn * block_N_2]). The FP16/BF16 path should keep its l_n buffer because T.tile.cast requires a UB destination whose dtype matches the output.

This removes one UB allocation and one vector copy. The performance impact may be small because the loss buffer is relatively small, but it also makes the dedicated FP32 path simpler and clearer.

File: examples/cross_entropy_loss/cross_entropy.py [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+
+    check_case(64, 131072, 128, 128)
+    check_case(256, 131072, 128, 128)
+
+    check_case(4, 32768, 128, 128)
+    check_case(4, 262144, 128, 128)
+
+    check_case(4, 131073, 128, 128)
+    check_case(4, 130000, 128, 128)
+
```
Comment (ShareableXue) [inline examples/cross_entropy_loss/cross_entropy.py]: The current suite contains several large-vocabulary cases, but the block configurations and target dtypes remain narrowly distributed, leaving important implementation branches and boundary conditions uncovered. would you please add more diverse cases?

File: examples/cross_entropy_loss/cross_entropy.py [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+                T.copy(y[bn * block_N_2], y_ub)
+                T.tile.ln(prev_sum, prev_sum)
+
+                for bc in T.serial(c_num):
+                    T.copy(x[bn * block_N_2, bc * block_C], x_32, pad_value=-T.infinity(CAL_DTYPE))
+
+                    T.tile.broadcast(max_2d, prev_max)
+                    T.tile.sub(x_32, x_32, max_2d)
+                    T.tile.broadcast(max_2d, prev_sum)
+                    T.tile.sub(x_32, x_32, max_2d)
```
Comment (ShareableXue) [inline examples/cross_entropy_loss/cross_entropy.py]: `prev_max` and `prev_sum` are invariant throughout the second `bc` loop, but the current implementation repeats two full 2D broadcasts for every class tile. With C=131072 and block_C=128, the same values are broadcast 1024 times, resulting in 2048 broadcast operations in total.

Could we consider moving both broadcasts outside the loop to avoid broadcasting the same values for every class tile? For example, we could introduce a separate logsum_2d buffer:

```python
T.tile.ln(prev_sum, prev_sum)
T.tile.broadcast(max_2d, prev_max, axis=1)
T.tile.broadcast(logsum_2d, prev_sum, axis=1)

for bc in T.serial(c_num):
    ...
    T.tile.sub(x_32, x_32, max_2d)
    T.tile.sub(x_32, x_32, logsum_2d)
```

For large C, this would reduce two broadcasts per iteration to only two broadcasts before the loop, potentially lowering vector-instruction and UB-access overhead.

One numerical consideration is that: 
Keeping the two-step subtraction would also be more numerically stable than combining the terms into lse = prev_max + log(prev_sum), where the smaller logarithmic term may be rounded away for large logits in FP32. It would be useful to evaluate this change together with the additional UB usage and on-device performance results.

**Review by ShareableXue — CHANGES_REQUESTED** [review]
Thanks for the optimization work! There are some promising improvements here. 

I found two issues that I think should be addressed before merging: the implicit broadcast-axis inference produces incorrect results for square tiles, and the advertised int64 target path currently fails to compile. 

I’ve also left a few other suggestions regarding test coverage and redundant broadcasts.

**Comment by ShareableXue** [comment]
good job! I have tested your Operator's performance, result as below:

**Performance tested:**  
- For shape `(1024, 1024, 128, 128)`, your kernel achieves **1.25×** the performance of the AscendC operator.  
- For shape `(4, 131072, 128, 128)`, your kernel is significantly slower than AscendC.

In general, this kernel works well for common shapes, but it does **not** perform well for small batch size with a large vocabulary dimension (small‑N / large‑C). We recommend adding a dedicated kernel branch for this case in a future patch.

**Suggestion:** Ready to upstream as is, with the above noted for follow‑up optimization.

**Review by LLMZhangYC — DISMISSED** [review]
lgtm

File: examples/cross_entropy_loss/example_cross_entro.py [hunk 7]
```diff
@@ -56,67 +57,172 @@ def main(
...
+                prev_sum = T.alloc_ub([block_N_2], CAL_DTYPE)
+                temp_exp = T.alloc_ub([block_N_2], CAL_DTYPE)
+                max_2d = T.alloc_ub([block_N_2, block_C], CAL_DTYPE)
+                logsum_2d = T.alloc_ub([block_N_2, block_C], CAL_DTYPE)
+
+                y_ub = T.alloc_ub([block_N_2], y_dtype)
+                l_n_32 = T.alloc_ub([block_N_2], CAL_DTYPE)
+                l_n = T.alloc_ub([block_N_2], x_dtype)
+
+                with T.Scope("V"):
```
Comment (LLMZhangYC) [inline examples/cross_entropy_loss/example_cross_entro.py]: Why not use combineCV and remove this scope

**Comment by ShareableXue** [comment]
Thanks for your PR! Please remove all Chinese comments in the code. You can either delete them entirely or translate them into English.

**Comment by ShareableXue** [comment]
And the benchmark results and the optimization path (tuning history) documentation should also be checked into the repository. Please place them under the operator directory as a separate Markdown (.md) file.

**Comment by ShareableXue** [comment]
Please provide more benchmark data for mainstream shapes, including pure kernel time and end-to-end time, and compare our implementation time with direct torch calls.

We must fully demonstrate that this operator meets the performance standards

**Review by fengz72 — APPROVED** [review]

File: examples/cross_entropy_loss/example_cross_entro.py (hunk 7 above)
Comment (zwj-sep) [reply examples/cross_entropy_loss/example_cross_entro.py]: 已在 3ba7481 启用 CombineCV pass 并移除手动 T.Scope("V")，验证通过。烦请重新 review/approve，谢谢！

**Comment by zwj-sep** [comment]
@ShareableXue 08-11 提出的 6 条问题已在 be615f1 全部修复（broadcast axis=1 显式指定、移除 int64、modulo 数据竞争改为线性 bn + 守卫、去掉 FP32 路径多余 buffer、循环外 broadcast、测试 19→23），烦请重新 review。@LLMZhangYC combineCV 建议已在 3ba7481 采纳（启用 CombineCV pass、移除手动 T.Scope），之前您的 approve 因新提交被 dismiss，也烦请重新确认。谢谢！

**Review by LLMZhangYC — APPROVED** [review]
approve
